File: 11.swift
import Foundation

class Monkey {
    var items: [UInt64] = [], destinations: [Int] = []
    var op: [String] = []
    var inspected: UInt64 = 0, div: UInt64
    var relief: Bool = true

    init(_ items: [UInt64], _ dest: [Int], _ op: [String], _ div: UInt64, _ relief: Bool) {
        self.items = items
        self.destinations = dest
        self.op = op
        self.div = div
        self.relief = relief
    }
    func inspect(_ ms: inout [Monkey]) {
        for item in items {
            let arg: UInt64 = op[7] == "old" ? item : UInt64(op[7])!
            var level: UInt64 = op[6] == "+" ? item + arg : item * arg
            if relief { level /= 3 }
            level %= lcm
            let to = level % self.div == 0 ? destinations[0] : destinations[1]
            ms[to].items.append(level)
        }
        inspected += UInt64(items.count)
        items = []
    }
}

var lcm: UInt64
var monkeys: [Monkey] = [], monkeys2: [Monkey] = []

while let line = readLine() {
    if line.hasPrefix("Monkey ") {
        let lines = [readLine(), readLine(), readLine(), readLine(), readLine()]
        let items = lines[0]!.numbers
        let dest = [Int(lines[3]!.numbers[0]), Int(lines[4]!.numbers[0])]
        let op = lines[1]!.components(separatedBy: [" "])
        let div = lines[2]!.numbers[0]
        monkeys.append(Monkey(items, dest, op, div, true))
        monkeys2.append(Monkey(items, dest, op, div, false))
   }
}
lcm = monkeys.map {$0.div}.reduce(1, *)
for _ in 0..<20 {
    for m in monkeys { m.inspect(&monkeys) }
}
for _ in 0..<10000 {
    for m in monkeys2 { m.inspect(&monkeys2) }
}
monkeys.sort { $1.inspected < $0.inspected }
monkeys2.sort { $1.inspected < $0.inspected }
print(monkeys[0].inspected * monkeys[1].inspected, monkeys2[0].inspected * monkeys2[1].inspected)

extension String {
    var numbers: [UInt64] {
        guard let regex = try? NSRegularExpression(
            pattern: "\\d+",
            options: [.caseInsensitive]) else {
            return []
        }
        let matches  = regex.matches(in: self, options: [],
            range: NSMakeRange(0, self.count))
        return matches.map { match in
            String(self[Range(match.range, in: self)!])
        }.map {
            UInt64($0)!
        }
    }
}

File: 11.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"regexp"
	"sort"
	"strconv"
	"strings"
)

type monkey struct {
	items []int64
	op    struct {
		name string
		arg  int64
		self bool
	}
	divisor   int
	dest      map[bool]int
	inspected int
}

func (m *monkey) inspect(monkeys []monkey, relief bool) {
	d := 1
	for _, mm := range monkeys {
		d *= mm.divisor
	}
	for _, item := range m.items {
		var level int64
		if m.op.self {
			level = item * item
		} else {
			switch m.op.name {
			case "*":
				level = item * m.op.arg
			case "+":
				level = item + m.op.arg
			}
		}
		if relief {
			level = level / 3
		}
		level = level % int64(d)
		var to = &monkeys[m.dest[level%int64(m.divisor) == 0]]
		to.items = append(to.items, level)
	}
	m.inspected += len(m.items)
	m.items = []int64{}
}

func business(monkeys []monkey) int {
	sort.Slice(monkeys, func(i, j int) bool {
		return monkeys[i].inspected > monkeys[j].inspected
	})
	return monkeys[0].inspected * monkeys[1].inspected
}

func main() {
	scanner := bufio.NewScanner(os.Stdin)
	monkeys := []monkey{}
	monkeys2 := []monkey{}
	for scanner.Scan() {
		if len(scanner.Text()) > 0 && scanner.Text()[:6] == "Monkey" {
			m := monkey{items: []int64{}, dest: map[bool]int{}}
			scanner.Scan()
			re := regexp.MustCompile("[0-9]+")
			for _, s := range re.FindAllString(scanner.Text(), -1) {
				n, _ := strconv.ParseInt(s, 0, 0)
				m.items = append(m.items, n)
			}
			scanner.Scan()
			ss := strings.Split(scanner.Text(), "=")
			var arg string
			fmt.Sscanf(ss[1], " old %s %s", &m.op.name, &arg)
			if arg == "old" {
				m.op.self = true
			} else {
				n, _ := strconv.ParseInt(arg, 0, 0)
				m.op.arg = n
			}
			scanner.Scan()
			fmt.Sscanf(scanner.Text(), "  Test: divisible by %d", &m.divisor)
			scanner.Scan()
			var to int
			fmt.Sscanf(scanner.Text(), "    If true: throw to monkey %d", &to)
			m.dest[true] = to
			scanner.Scan()
			fmt.Sscanf(scanner.Text(), "    If false: throw to monkey %d", &to)
			m.dest[false] = to
			monkeys = append(monkeys, m)
			monkeys2 = append(monkeys2, m)
		}
	}
	for i := 0; i < 20; i++ {
		for m := 0; m < len(monkeys); m++ {
			(&monkeys[m]).inspect(monkeys, true)
		}
	}
	for i := 0; i < 10000; i++ {
		for m := 0; m < len(monkeys2); m++ {
			(&monkeys2[m]).inspect(monkeys2, false)
		}
	}
	fmt.Println(business(monkeys), business(monkeys2))
}
